src/cart/crt_context.c

```diff
@@ -173,7 +173,42 @@ crt_context_init(crt_context_t crt_ctx)
 }
 
 int
-crt_context_provider_create(crt_context_t *crt_ctx, int provider)
+crt_context_provider_create(crt_context_t *crt_ctx, crt_provider_t provider, bool primary);
+
+int
+crt_context_create_on_provider(crt_context_t *crt_ctx, const char *provider, bool primary)
+{
+	int	provider_idx = -1;
+
+	provider_idx = crt_str_to_provider(provider);
+	if (provider_idx == -1) {
+		D_ERROR("Invalid requested provider '%s'\n", provider);
+		return -DER_INVAL;
+	}
+
+	return crt_context_provider_create(crt_ctx, provider_idx, primary);
+}
+
+int
+crt_context_uri_get(crt_context_t crt_ctx, char **uri)
+{
+	struct crt_context	*ctx = NULL;
+
+	if (crt_ctx == NULL || uri == NULL) {
+		D_ERROR("Invalid null parameters\n");
+		return -DER_INVAL;
+	}
+
+	ctx = crt_ctx;
+	D_STRNDUP(*uri, ctx->cc_self_uri, CRT_ADDR_STR_MAX_LEN);
+	if (*uri == NULL)
+		return DER_NOMEM;
```

Comment:
@frostedcmos shouldn't this be -DER_NOMEM?

Reply:
yep good catch. will patch up in the next pr